upf/upf_pfcp_api.c

```diff
@@ -1041,27 +1246,89 @@ handle_update_pdr (upf_session_t * sx, pfcp_update_pdr_t * update_pdr,
       {
-	update->pdi.fields |= F_PDI_LOCAL_F_TEID;
-	/* TODO validate TEID and mask */
-	update->pdi.teid = pdr->pdi.f_teid;
+        update->pdi.fields |= F_PDI_UE_IP_ADDR;
+        update->pdi.ue_addr = pdr->pdi.ue_ip_address;
+
+        if (!ISSET_BIT (pdr->pdi.grp.fields, PDI_SDF_FILTER) &&
+            !ISSET_BIT (pdr->pdi.grp.fields, PDI_APPLICATION_ID))
+          {
+            /* neither SDF, nor Application Id, generate a wildcard
+               ACL to make ACL scanning simpler */
+            update->pdi.fields |= F_PDI_SDF_FILTER;
+            vec_reset_length (update->pdi.acl);
+            vec_add1 (update->pdi.acl, wildcard_acl);
+          }
       }
 
-    if (ISSET_BIT (pdr->pdi.grp.fields, PDI_UE_IP_ADDRESS))
+    p = hash_get_mem (gtm->nwi_index_by_name, nwi->name);
+    if (!p)
       {
-	update->pdi.fields |= F_PDI_UE_IP_ADDR;
-	update->pdi.ue_addr = pdr->pdi.ue_ip_address;
+        upf_debug("Can't find nwi idx by name\n");
+        r = -1;
+        break;
+      }
 
-	if (!ISSET_BIT (pdr->pdi.grp.fields, PDI_SDF_FILTER) &&
-	    !ISSET_BIT (pdr->pdi.grp.fields, PDI_APPLICATION_ID))
-	  {
-	    /* neither SDF, nor Application Id, generate a wildcard
-	       ACL to make ACL scanning simpler */
-	    update->pdi.fields |= F_PDI_SDF_FILTER;
-	    vec_reset_length (update->pdi.acl);
-	    vec_add1 (update->pdi.acl, wildcard_acl);
-	  }
+    idx = p[0];
+    res = &gtm->upip_res[idx];
+
+
+    if (ISSET_BIT (pdr->pdi.grp.fields, PDI_F_TEID))
+      {
+        u32 teid = 0;
+        bool chosen = false;
+
+        update->pdi.fields |= F_PDI_LOCAL_F_TEID;
+        // We only generate F_TEID if NWI is defined
+        if ((pdr->pdi.f_teid.flags & F_TEID_CH) && res)
+          {
+            if ((pdr->pdi.f_teid.flags & F_TEID_CHID))
+              {
+                u8 choose_id = pdr->pdi.f_teid.choose_id;
+                teid = teid_lookup_by_choose_id(sx, choose_id);
+                if (teid)
+                  chosen = true;
+              }
+            if (!chosen)
+              {
+                teid = process_teid_generation(gtm, pdr->pdi.f_teid.choose_id,
+                                              pdr->pdi.f_teid.flags, res, sx);
+                if (!teid)
+                  {
+                    upf_debug("Can't generate TEID for pdr=%d\n", pdr->pdr_id);
+                    r = -1;
+                    break;
+                  }
+              }
+            update->pdi.teid.teid = teid;
+            if ((!is_zero_ip4_address (&res->ip4)) && (pdr->pdi.f_teid.flags && F_TEID_V4))
+              {
+                update->pdi.teid.ip4 = res->ip4;
+                update->pdi.teid.flags |= F_TEID_V4;
+              }
+            if ((!is_zero_ip6_address (&res->ip6)) && (pdr->pdi.f_teid.flags && F_TEID_V4))
+              {
+                update->pdi.teid.flags |= F_TEID_V6;
+                update->pdi.teid.ip6 = res->ip6;
+              }
+            pfcp_created_pdr_t *created_pdr;
```

Comment:
I suppose it's better to move this on top of the code block with F_TEID field checks. This is only needed if we are filling up response with created_pdrs if we done teid generation on UP side. This will be unused in all other situations